### PR TITLE
👷 Remove escrow on inflationary rewards

### PR DESCRIPTION
**Pull request by Flocqst**

### CODE REVIEW

**Review by JaredBorders — CHANGES_REQUESTED** [review]

File: contracts/StakingRewardsV2.sol [hunk 1]
```diff
@@ -213,19 +214,27 @@ contract StakingRewardsV2 is
     ///////////////////////////////////////////////////////////////*/
 
     /// @inheritdoc IStakingRewardsV2
-    function stake(uint256 _amount) external whenNotPaused updateReward(msg.sender) {
+    function stake(uint256 _amount) external whenNotPaused {
```
[the remark this answers is not in the document]
Comment (Flocqst) [reply contracts/StakingRewardsV2.sol:217]: Updated

File: contracts/StakingRewardsV2.sol [hunk 2]
```diff
@@ -213,19 +214,27 @@ contract StakingRewardsV2 is
...
+        _stake(msg.sender, _amount);
+
+        // transfer token to this contract from the caller
+        kwenta.transferFrom(msg.sender, address(this), _amount);
+    }
+
+    function _stake(address _account, uint256 _amount)
+        internal
+        whenNotPaused
+        updateReward(msg.sender)
```
[the remark this answers is not in the document]
Comment (Flocqst) [reply contracts/StakingRewardsV2.sol:227]: You're right, previously was `msg.sender` but now that `_stake` is called when calling `compoundOnBehalf`, it needs to be account otherwise in this specific case it's not appropriate.

Will improve the `compoundOnBehalf` test to catch this.

File: contracts/StakingRewardsV2.sol [hunk 3]
```diff
@@ -343,10 +352,25 @@ contract StakingRewardsV2 is
             // emit reward claimed event and index account
             emit RewardPaid(_account, reward);
 
-            // transfer token from this contract to the rewardEscrow
-            // and create a vesting entry at the _to address
-            kwenta.transfer(address(rewardEscrow), reward);
-            rewardEscrow.appendVestingEntry(_to, reward);
+            // transfer token from this contract to the account
+            // as newly issued rewards from inflation are now issued as non-escrowed
+            kwenta.transfer(_to, reward);
```
[the remark this answers is not in the document]
Comment (Flocqst) [reply contracts/StakingRewardsV2.sol:357]: There is the compound function if you want to do so, this is just for claiming rewards, so i would keep it this way.

File: contracts/StakingRewardsV2.sol [hunk 4]
```diff
@@ -343,10 +352,25 @@ contract StakingRewardsV2 is
...
+        }
+    }
+
+    function _getRewardCompounding(address _account)
```
[the remark this answers is not in the document]
Comment (Flocqst) [reply contracts/StakingRewardsV2.sol:361]: This is called when compounding, basically it is doing the same as `_getReward` but without transfering because the rewards are staked right after this (can be staked immediatly after calling this (see `_compound()`).

Added natspec to specify this.

**Comment by Flocqst** [comment]
I've merged the 'Staking Cooldown Removal' feature into this branch and also pulled the USDC rewards from main branch. This PR should be the one that goes live with the KSX deployment.

File: contracts/RewardEscrowV2.sol [hunk 5]
```diff
@@ -19,7 +19,7 @@ import {IEscrowMigrator} from "./interfaces/IEscrowMigrator.sol";
 /// @title KWENTA Reward Escrow V2
 /// @author Originally inspired by SYNTHETIX RewardEscrow
 /// @author Kwenta's RewardEscrow V1 by JaredBorders (jaredborders@proton.me), JChiaramonte7 (jeremy@bytecode.llc)
-/// @author RewardEscrowV2 by tommyrharper (tom@zkconsulting.xyz)
+/// @author RewardEscrowV2 by tommyrharper (tom@zkconsulting.xyz), Flocqst (florian@kwenta.io)
```
Comment (JaredBorders) [inline contracts/RewardEscrowV2.sol:22]: nit: I would break each author up on separate lines

**Review by JaredBorders — APPROVED** [review]
Everything LGTM! Nice work

File: contracts/RewardEscrowV2.sol [hunk 6]
```diff
@@ -379,7 +379,7 @@ contract RewardEscrowV2 is
                 unchecked {
                     amountToUnstake = totalWithFee - unstakedEscrow;
                 }
-                stakingRewards.unstakeEscrowSkipCooldown(msg.sender, amountToUnstake);
+                stakingRewards.unstakeEscrowAdmin(msg.sender, amountToUnstake);
```
Comment (etn0m) [inline contracts/RewardEscrowV2.sol:382]: Is this necessary if there's no cooldown? Simply call unstakeEscrow no?

File: contracts/StakingRewardsV2.sol [hunk 7]
```diff
@@ -239,27 +218,34 @@ contract StakingRewardsV2 is
...
+    function _stake(address _account, uint256 _amount)
+        internal
+        whenNotPaused
+        updateReward(_account)
+    {
         if (_amount == 0) revert AmountZero();
 
         // update state
-        userLastStakeTime[msg.sender] = block.timestamp;
+        userLastStakeTime[_account] = block.timestamp;
```
Comment (etn0m) [inline contracts/StakingRewardsV2.sol]: Why do we still keep track of userLastStakeTime?

File: contracts/StakingRewardsV2.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (etn0m) [reply contracts/StakingRewardsV2.sol]: Hmm, is there a difference to this vs. the code in _getReward other than transferring? I see duplicate code. I was thinking that this reward calculating logic would be shared between the two and getReward only has the distinction of transferring to the user.

ie.
```solidity
getReward() {
    (kwenta, usdc) = _calculateReward()
    transfer kwenta
    transfer USDC
}

compound() {
    (kwenta, usdc) = _calculateReward()
    stake kwenta
    transfer USDC
}

_calculateReward() {
...
}
```

TLDR; we should be practicing [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) here to minimize room for error.

**Review by jcmonte — CHANGES_REQUESTED** [review]

File: contracts/RewardEscrowV2.sol (hunk 6 above)
Comment (Flocqst) [reply contracts/RewardEscrowV2.sol:382]: You have to call unstakeEscrowAdmin otherwise `msg.sender` for
```
function unstakeEscrow(uint256 _amount) external {
        _unstakeEscrow(msg.sender, _amount);
    }
```

would be the rewardEscrowV2 contract.

File: contracts/StakingRewardsV2.sol (hunk 7 above)
Comment (Flocqst) [reply contracts/StakingRewardsV2.sol]: We don't anymore you're right, removed it.

File: contracts/StakingRewardsV2.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (Flocqst) [reply contracts/StakingRewardsV2.sol]: Refactored with a `_processReward` so that there is no duplicate logic

File: contracts/StakingRewardsV2.sol [hunk 8]
```diff
@@ -345,61 +338,43 @@ contract StakingRewardsV2 is
...
+    /// @param _account The address of the account to process rewards for
+    /// @param _to The address to transfer rewards to
+    /// @param transferKwenta Boolean flag to determine if Kwenta should be transferred
+    /// @return kwentaReward The amount of Kwenta reward processed
+    function _processReward(address _account, address _to, bool transferKwenta)
         internal
-        whenNotPaused
-        updateReward(_account)
-        returns (uint256 reward)
+        returns (uint256 kwentaReward)
```
Comment (etn0m) [inline contracts/StakingRewardsV2.sol:353]: Doesn't seem you're using kwentaReward outside of _processReward, but for consistency sake you should either return both kwenta and USDC reward as a tuple or return neither.

File: contracts/StakingRewardsV2.sol [hunk 9]
```diff
@@ -81,9 +75,6 @@ contract StakingRewardsV2 is
     /// @notice summation of rewardRate divided by total staked tokens
     uint256 public rewardPerTokenStored;
 
-    /// @inheritdoc IStakingRewardsV2
-    uint256 public cooldownPeriod;
```
Comment (cmontecoding) [inline contracts/StakingRewardsV2.sol:85]: are we redeploying staking v2 or upgrading? because im pretty sure if we upgrade this removal will brick everything

Comment (cmontecoding) [inline contracts/StakingRewardsV2.sol:85]: are we redeploying staking v2 or upgrading? because im pretty sure if we upgrade this removal will brick everything

File: contracts/StakingRewardsV2.sol [hunk 10]
```diff
@@ -81,9 +75,6 @@ contract StakingRewardsV2 is
...
-
     /// @notice represents the rewardPerToken
     /// value the last time the staker calculated earned() rewards
     mapping(address => uint256) public userRewardPerTokenPaid;
```
Comment (cmontecoding) [inline contracts/StakingRewardsV2.sol:80]: it wont let me comment a couple lines below for some reason but ``mapping(address => uint256) public userLastStakeTime;`` should be removed because we remove the use of it everywhere in the code

File: contracts/StakingRewardsV2.sol [hunk 11]
```diff
@@ -361,31 +338,43 @@ contract StakingRewardsV2 is
...
+    {
+        // Process Kwenta reward
+        kwentaReward = rewards[_account];
+        if (kwentaReward > 0) {
             // update state (first)
             rewards[_account] = 0;
 
             // emit reward claimed event and index account
-            emit RewardPaid(_account, reward);
+            emit RewardPaid(_account, kwentaReward);
```
Comment (cmontecoding) [inline contracts/StakingRewardsV2.sol:362]: the other event is called RewardPaidUSDC

this should be called RewardPaidKwenta to match

File: contracts/StakingRewardsV2.sol [hunk 12]
```diff
@@ -361,31 +338,43 @@ contract StakingRewardsV2 is
...
+    /// @return kwentaReward The amount of Kwenta reward processed
+    function _processReward(address _account, address _to, bool transferKwenta)
+        internal
+        returns (uint256 kwentaReward)
+    {
+        // Process Kwenta reward
+        kwentaReward = rewards[_account];
+        if (kwentaReward > 0) {
             // update state (first)
             rewards[_account] = 0;
```
Comment (cmontecoding) [inline contracts/StakingRewardsV2.sol:359]: if transferKwenta is set to false meaning they want to compound their rewards then doesnt this line ruin that?

their rewards are set to 0 but they dont get transferred their kwenta, so next time this is called again (im assuming after they waited for it to compound in this contract), then their rewards will be 0 or close to 0 and therefore they lost their kwenta entirely?

if im understanding this correctly then this ``rewards[_account] = 0;`` line shouldnt apply if transferKwenta == false

File: contracts/StakingRewardsV2.sol [hunk 13]
```diff
@@ -397,8 +386,8 @@ contract StakingRewardsV2 is
     /// @dev internal helper to compound for a given account
     /// @param _account the account to compound for
     function _compound(address _account) internal {
-        _getReward(_account);
-        _stakeEscrow(_account, unstakedEscrowedBalanceOf(_account));
+        uint256 reward = _processReward(_account, _account, false);
```
Comment (cmontecoding) [inline contracts/StakingRewardsV2.sol:390]: will reward be 0 here?

File: contracts/StakingRewardsV2.sol (hunk 12 above)
Comment (cmontecoding) [reply contracts/StakingRewardsV2.sol:359]: oh upon further look, kwentaReward is extrapolating this value? and it hopefully gets relogged when calling _stake in _compound?

**Review by cmontecoding — COMMENTED** [review]
initial review of main contracts
looks pretty straight forward so far

File: test/foundry/unit/StakingRewardsV2/StakingRewardsV2OnBehalfActionsTests.t.sol [hunk 14]
```diff
@@ -175,76 +175,6 @@ contract StakingRewardsV2OnBehalfActionsTests is DefaultStakingV2Setup {
         stakingRewardsV2.approveOperator(owner, true);
     }
 
-    /*//////////////////////////////////////////////////////////////
-                        Get Reward On Behalf
-    //////////////////////////////////////////////////////////////*/
-
-    function test_getRewardOnBehalf() public {
```
Comment (cmontecoding) [inline test/foundry/unit/StakingRewardsV2/StakingRewardsV2OnBehalfActionsTests.t.sol:182]: why do we remove this test and the accompanying fuzz test?

File: test/foundry/unit/StakingRewardsV2/StakingRewardsV2OnBehalfActionsTests.t.sol [hunk 15]
```diff
@@ -356,18 +286,20 @@ contract StakingRewardsV2OnBehalfActionsTests is DefaultStakingV2Setup {
     }
 
     /*//////////////////////////////////////////////////////////////
-                    Get Reward And Stake On Behalf
+                    Get Reward On Behalf
     //////////////////////////////////////////////////////////////*/
 
-    function test_Get_Reward_And_Stake_On_Behalf() public {
+    function test_Get_Reward_On_Behalf() public {
```
Comment (cmontecoding) [inline test/foundry/unit/StakingRewardsV2/StakingRewardsV2OnBehalfActionsTests.t.sol:292]: its good that we are testing getReward() by itself now (im assuming we didnt before) but why are we just replacing the "on behalf" test? we no longer test getting the reward AND staking anymore

we should have both this test & the test_Get_Reward_And_Stake. not either or

Comment (cmontecoding) [reply test/foundry/unit/StakingRewardsV2/StakingRewardsV2OnBehalfActionsTests.t.sol:292]: this reminds me: where do we test getReward (at least the new changes)? cant find it

File: test/foundry/unit/StakingRewardsV2/StakingV2Checkpointing.t.sol [hunk 16]
```diff
@@ -354,7 +336,7 @@ contract StakingV2CheckpointingTests is DefaultStakingV2Setup {
             assertEq(value, previousTotal + amountToStake);
 
             // move beyond cold period
-            vm.warp(block.timestamp + stakingRewardsV2.cooldownPeriod());
+            vm.warp(block.timestamp + 2 weeks);
```
Comment (cmontecoding) [inline test/foundry/unit/StakingRewardsV2/StakingV2Checkpointing.t.sol:339]: how come we mostly remove the cooldown period in this test file but then sometimes we add back this warp forward 2 weeks?

File: test/foundry/unit/StakingRewardsV2/StakingV2RewardCalculations.t.sol [hunk 17]
```diff
@@ -513,10 +513,10 @@ contract StakingV2RewardCalculationTests is DefaultStakingV2Setup {
             getStakingRewardsV2(user1);
 
             // check rewards
-            rewards = rewardEscrowV2.escrowedBalanceOf(user1);
+            balance = kwenta.balanceOf(user1);
+            assertEq(balance, expectedRewards);
             rewardsUsdc = usdc.balanceOf(user1);
-            assertEq(rewards, expectedRewards);
-            assertApproxEqAbs(rewardsUsdc, expectedUsdcRewards, 10);
+            assertApproxEqAbs(rewardsUsdc, expectedUsdcRewards, 50);
```
Comment (cmontecoding) [inline test/foundry/unit/StakingRewardsV2/StakingV2RewardCalculations.t.sol:519]: why do we change this from 10 to 50? does that make the assert less accurate?

i just tested this locally with 10 and it still passes

File: test/foundry/integration/stakingV2.upgrade.t.sol [hunk 18]
```diff
@@ -232,14 +232,16 @@ contract StakingV2UpgradeTests is DefaultStakingV2Setup {
...
+        assertEq(1 ether, rewardEscrowV2.escrowedBalanceOf(user1));
+        assertEq(1 ether, rewardEscrowV2.unstakedEscrowedBalanceOf(user1));
 
         // stake the rewards
         stakeAllUnstakedEscrowV2(user1);
+        stakeFundsV2(user1, 1 weeks);
         assertEq(2 ether + 1 weeks, stakingRewardsV2.balanceOf(user1));
-        assertEq(1 ether + 1 weeks, stakingRewardsV2.escrowedBalanceOf(user1));
+        assertEq(1 ether, stakingRewardsV2.escrowedBalanceOf(user1));
         assertEq(0, rewardEscrowV2.unstakedEscrowedBalanceOf(user1));
```
Comment (cmontecoding) [inline test/foundry/integration/stakingV2.upgrade.t.sol:245]: this test suite needs to also test that all the accessible storage slots are not bricked

**Review by cmontecoding — CHANGES_REQUESTED** [review]
just reviewed all the tests. aside: i cant find where we explicitly test these new _processReward changes. 

important: we need to test the upgrade from a current fork --> new deployment (much like how cc-vesting does it). this should be done before we upgrade to ensure everything will be good, especially storage slots.
this may be doable in stakingV2.upgrade.t.sol. if not then we might have to add a new file. this may be difficult because i think everthing (testing and upgrading) is baked together and deeply webbed
